src/dataflow/src/render/reduce.rs

```diff
@@ -609,14 +619,27 @@ where
 fn build_hierarchical<G>(
     collection: Collection<G, (Row, Row)>,
     aggr: &AggregateFunc,
+    expected_group_size: Option<usize>,
 ) -> Collection<G, (Row, Row)>
 where
     G: Scope,
     G::Timestamp: Lattice,
 {
+    let mut shifts = vec![];
+    let mut current = 4u64;
+
+    let limit = expected_group_size.unwrap_or(1_000_000);
```

Comment:
This is the only thing I'm not certain about. Some fraction of the use cases are going to be keyless aggregations; if they get sketchy performance on e.g. 100M records, until they read the docs about improving that, that is possibly worse than the memory saving other get from the lower defaults? I'm not sure the right way to reason this out tbh; whether we want to provide good worst case behavior by default, or good average case behavior (each at the expense of the other).

Reply:
yeah I had the same reservation.

One thing I noticed though is that when I tried a keyless aggregate we seemed to hold memory proportional to the number of distinct values and not the total number of values (point still taken that someone taking the min of 100M distinct values will be unhappy).

Perhaps one thing I could try to do is quantify how unhappy a user with 100M rows will be in the current scheme? That could help us make this decision maybe

Reply:
We talked offline and agreed to bump the threshold to 4B instead of 1 million in the case where theres no hint